### PR TITLE
[Ahuna] Allow moving an assistant back to private

### LINKED ISSUE #3270: Allow moving an assistant back to private
This is backend work for team sharing parts of this climb
The route `pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts` currently does not support moving an assistant back to private. 

~We should remove the assistant from everybody's list except the person moving it back to private~ => after discussion, we leave the assistant in users' lists. If scope is reverted to published/workspace, users get back to their setting automatically

- ensure when moving back to private that it has a listOverride "in-list" in the AgentUserRelation table
- And double-check that for private assistants, no option to "remove from list" exist

### PR DESCRIPTION
**Pull request by philipperolet**
## Description
Fixes #3270

API route to change scope of an assistant allows private, with associated logic (see issue)

## Risk
On failure mid-moving back to private (e.g. failed after removing from list but before changing scope)
This needs not be adressed (logs / notifications are enough)

### CODE REVIEW

File: front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts [hunk 1]
```diff
@@ -86,17 +90,63 @@ async function handler(
...
         assistant.scope !== "private" &&
         bodyValidation.right.scope === "private"
       ) {
-        return apiError(req, res, {
-          status_code: 400,
-          api_error: {
-            type: "invalid_request_error",
-            message: "Non-private assistants cannot be set back to private.",
-          },
+        // switching an assistant back to private: removing it from all user's
```
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts:98]: all users' list or all the user's list?

Comment (philipperolet) [reply front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts:98]: indeed, `all users' lists` was the right wording

File: front/lib/api/assistant/user_relation.ts [hunk 2]
```diff
@@ -112,3 +112,44 @@ export async function setAgentUserListstatus({
...
+  if (!userId || !workspaceId || !auth.isUser())
+    return new Err(new Error("User or workspace not found"));
+
+  if (agentConfiguration.status !== "active") {
+    return new Err(new Error("Agent is not active"));
+  }
+
+  const statusesDeleted = await AgentUserRelation.destroy({
+    where: {
+      userId,
```
Comment (spolu) [inline front/lib/api/assistant/user_relation.ts:145]: Re the other comment, then I don't understand this code. It removes the user relation not all users' relations ?

**Comment by philipperolet** [comment]
After quick chat with ed, turns out we're better leaving the agent in users' lists, notably in case user reverts

**Comment by spolu** [comment]
Make sense indeed IMHO

File: front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts [hunk 3]
```diff
@@ -86,25 +85,63 @@ async function handler(
...
+            api_error: {
+              type: "internal_server_error",
+              message: setRes.error.message,
+            },
+          });
+        }
       }
 
-      const result = await setAgentScope(
+      const result = await createOrUpgradeAgentConfiguration(
```
Comment (spolu) [inline front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts:134]: We def don't want to create a new version for that do we?

**Review by spolu — COMMENTED** [review]
Not sure I see why we craete a new full version?

**Comment by philipperolet** [comment]
As discussed IRL, we create a new version:
- to be consistent to when 'just the scope' is changed with assistant builder
- to have a log of who made the change -- changing scope is ~an edit just like any other assistant edit
- so that the person that sets it private becomes the last author, and as such the one to see it in "private"

File: front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts (hunk 3 above)
Comment (fontanierh) [reply front/pages/api/w/[wId]/assistant/agent_configurations/[aId]/scope.ts:134]: As per IRL, I understand the reasoning behind the change and I am OK with it.
We could argue that the "scope" isn't really part of an agent configuration and is more a "metadata" so creating an entire new version is def a bit heavy, but consistency with assistant builder is a strong argument here 👍

**Review by fontanierh — APPROVED** [review]

**Pull request merged**